include/mlir/Dialect/AffineOps/AffineOps.td

```diff
@@ -260,6 +260,82 @@ def AffineMinOp : Affine_Op<"min"> {
   let hasFolder = 1;
 }
 
+def AffinePrefetchOp : Affine_Op<"prefetch"> {
+  let summary = "affine prefetch operation";
+  let description = [{
+    The "affine.prefetch" op prefetches data from a memref location described
+    with an affine subscript similar to affine.load. For example:
+
+      affine.prefetch %0[%i, %j + 5], %w, %h, %d : memref<400x400xi32>
+
+    %w is true if the fetch should be for a write, false for a read, %h is a
+    temporal locality specifier ranging from (0) - no locality, to (3) -
+    extremely local keep in cache. %d specifies whether the prefetch
+    is performed on the data cache, false if on the instruction cache.
+  }];
+
+  let arguments = (ins AnyMemRef:$memref, Variadic<Index>:$indices,
+                   BoolAttr:$isWrite,
+                   Confined<I32Attr, [IntMinValue<0>,
+                     IntMaxValue<3>]>:$localityHint,
+                   BoolAttr:$isDataCache);
+
+  let builders = [OpBuilder<
+    "Builder *builder, OperationState &result, Value *memref,"
+    "AffineMap map, ArrayRef<Value *> mapOperands, bool isWrite,"
```

Comment:
Any pointers on how I could add this to the ODS?

Reply:
> builder.create<AffinePrefetchOp>(..., AffinePrefetchOp::Read, AffinePrefetchOp::MaximumLocalityHint, AffinePrefetchOp::DataCache)

The above really buys little over the current:

builder.create<AffinePrefetchOp>(..., /*isRead=*/false, /*localityHint=*/..., /*isDataCache=*/true);

assuming that checkers can catch out of order arguments. In fact, I find the latter more readable.

Reply:
> Any pointers on how I could add this to the ODS?

Since we ultimately generate C++ from ODS, you can declare classes enums in the header file in which you include the ODS-generated code, then just use them in builder signature.

> assuming that checkers can catch out of order arguments. In fact, I find the latter more readable.

Readability is subjective so I won't insist on it. However, if I can have the mandatory tool (compiler) do the check, I take it instead of relying on some non-mandatory tool (linter) to complain that somebody forgot a comment or used a mismatching comment. A typical problem with sequences of non-typed bools/ints is the API change that changes, e.g., the order of arguments without breaking the compilation. Yes, we have linters that can catch it, but it is more a bandaid for legacy codebases IMO since we now have language features (enum classes) that prevent this problem from appearing in the first place.

Reply:
Okay. I prefer leaving this as is.